packages/eas-cli/src/update/utils.ts

```diff
@@ -269,3 +274,90 @@ export function getBranchDescription(branch: UpdateBranchFragment): FormattedBra
     },
   };
 }
+
+export function isBundleDiffingEnabled(exp: ExpoConfig): boolean {
+  return (exp.updates as any)?.enableBsdiffPatchSupport === true;
+}
+
+// Make authenticated requests to the launch asset URL with diffing headers
+export async function prewarmDiffingAsync(
+  graphqlClient: ExpoGraphqlClient,
+  appId: string,
+  newUpdates: UpdatePublishMutation['updateBranch']['publishUpdateGroups']
+): Promise<void> {
+  const DUMMY_EMBEDDED_UPDATE_ID = '00000000-0000-0000-0000-000000000000';
+
+  const toPrewarm = [] as {
+    update: UpdatePublishMutation['updateBranch']['publishUpdateGroups'][0];
+    launchAssetKey: string;
+  }[];
+
+  for (const update of newUpdates) {
+    try {
+      const manifest = JSON.parse(update.manifestFragment);
+      const launchAssetKey: string | undefined = manifest.launchAsset?.storageKey;
+      const requestedUpdateId: string = update.id;
+      if (!launchAssetKey || !requestedUpdateId) {
+        continue;
+      }
+      toPrewarm.push({
+        update,
+        launchAssetKey,
+      });
+    } catch {
+      // ignore updates without launch assets
+    }
+  }
+
+  await Promise.allSettled(
+    toPrewarm.map(async ({ update, launchAssetKey }) => {
+      try {
+        // Check to see if there's a second most recent update so we can pre-emptively generate a patch for it
+        const updatePublishPlatform = update.platform as UpdatePublishPlatform;
+        const updateIds = await BranchQuery.getUpdateIdsOnBranchAsync(graphqlClient, {
+          appId,
+          branchName: update.branch.name,
+          platform: updatePublishPlatformToAppPlatform[updatePublishPlatform],
+          runtimeVersion: update.runtimeVersion,
+          limit: 2,
+        });
+        if (updateIds.length !== 2) {
+          return;
+        }
+        const nextMostRecentUpdateId = updateIds[1];
+
+        const signed = await AssetQuery.getSignedUrlsAsync(graphqlClient, update.id, [
+          launchAssetKey,
+        ]);
+        const first = signed?.[0];
+        if (!first?.url || !first?.headers) {
+          return;
+        }
+
+        const headers: Record<string, string> = {
+          ...(first.headers as Record<string, string>),
```

Comment:
I think URL is required? Or is this checking empty string?

Also since headers is optional I think we should continue with the fetch even if headers is required.


```suggestion
        if (!first) {
          return;
        }

        const headers: Record<string, string> = {
          ...(first.headers as Record<string, string> | undefined),
```